pkg/webhook/workspace/handler/metadata.go

```diff
@@ -0,0 +1,101 @@
+//
+// Copyright (c) 2019-2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+package handler
+
+import (
+	"fmt"
+	"github.com/che-incubator/che-workspace-operator/pkg/controller/workspace/model"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (h *WebhookHandler) mutateMetadataOnCreate(o *metav1.ObjectMeta) error {
+	if o.GetAnnotations() == nil {
+		return fmt.Errorf("annotations are missing while %s is required", model.WorkspaceCreatorAnnotation)
+	} else {
+		if _, ok := o.GetAnnotations()[model.WorkspaceCreatorAnnotation]; !ok {
+			return fmt.Errorf("'%s' annotation is missing", model.WorkspaceCreatorAnnotation)
+		}
+	}
+
+	if o.GetLabels() == nil {
+		return fmt.Errorf("labels are missing while %s is required", model.WorkspaceIDLabel)
+	} else {
+		if _, ok := o.GetLabels()[model.WorkspaceIDLabel]; !ok {
+			return fmt.Errorf("'%s' label is missing", model.WorkspaceIDLabel)
+		}
+	}
+
+	return nil
+}
+
+func (h *WebhookHandler) mutateMetadataOnUpdate(oldMeta, newMeta *metav1.ObjectMeta) (bool, error) {
+	if oldMeta.Annotations == nil {
+		oldMeta.Annotations = map[string]string{}
+	}
+	if newMeta.Annotations == nil {
+		newMeta.Annotations = map[string]string{}
+	}
+	updatedAnnotations, err := mutateAnnotationsOnUpdate(oldMeta.Annotations, newMeta.Annotations)
+	if err != nil {
+		return false, err
+	}
+
+	if oldMeta.Labels == nil {
+		oldMeta.Labels = map[string]string{}
+	}
+	if newMeta.Labels == nil {
+		newMeta.Labels = map[string]string{}
+	}
+	updatedLabels, err := mutateLabelsOnUpdate(oldMeta.Labels, newMeta.Labels)
+	if err != nil {
+		return false, err
+	}
+
+	return updatedAnnotations || updatedLabels, nil
+}
+
+func mutateAnnotationsOnUpdate(old, new map[string]string) (bool, error) {
+	oldCreator, found := old[model.WorkspaceCreatorAnnotation]
+	if !found {
+		return false, fmt.Errorf("'%s' annotation is required. Update Controller and restart your workspace", model.WorkspaceCreatorAnnotation)
+	}
+
+	newCreator, found := new[model.WorkspaceCreatorAnnotation]
+	if !found {
+		new[model.WorkspaceCreatorAnnotation] = oldCreator
```

Comment:
Is this the correct logic? It's more permissive but harder to reason about -- if my update to the workspace removes the owner annotation, I'm not sure we should just add it and pretend nothing is wrong. I'm not 100% sure here though.

Reply:
I implemented this logic by keeping in mind that user is able to do `kubectl apply -f ...`  where original file does not contain creator annotations.
But since we don't really want user to modify that deployments manually, I'm OK with making webhook for pod and deployment validating and just throwing error when creator or workspace id is missing, WDYT?

Reply:
Ah we definitely need to support `kubectl apply -f`. In that scenario though, does the `old` workspace not have the creator annotation? I figure it should be set upon creation, so even if I `apply` a new file over it, the old annotation will always be there. 

If `oldWksp` does not have an annotation, I can claim your workspace by grabbing your workspace yaml from the cluster, making a no-op modification, and applying it.

Reply:
Ah you're right, `kubectl apply -f` is probably a case we need to support. Ignore this comment :)

Reply:
Ah you're right, `kubectl apply -f` is probably a case we need to support. Ignore this comment :)